python/glow/hail/functions.py

```diff
@@ -0,0 +1,183 @@
+            f"zip_with({sample_id_expr}, entries, (s, e) -> named_struct({struct_expr}))")
+    else:
+        struct_expr = ' ,'.join(base_struct_args)
+        genotypes_col = fx.expr(f"transform(entries, e -> named_struct({struct_expr}))")
+    genotypes_col = genotypes_col.alias("genotypes")
+
+    assert check_return_type(genotypes_col)
+    return genotypes_col
+
+
+def __get_base_cols(row: StructExpression) -> List[Column]:
+    assert check_argument_types()
+
+    contig_name_col = fx.col("`locus.contig`").alias("contigName")
+
+    start_col = (fx.col("`locus.position`") - 1).cast("long").alias("start")
+
+    end_col = start_col + fx.length(fx.element_at("alleles", 1))
+    has_info = 'info' in row and isinstance(row.info.dtype, tstruct)
+    if has_info and 'END' in row.info and row.info.END.dtype == tint:
+        end_col = fx.coalesce(fx.col("`info.END`"), end_col)
+    end_col = end_col.cast("long").alias("end")
+
+    names_elems = []
+    if 'varid' in row and row.varid.dtype == tstr:
+        names_elems.append("varid")
+    if 'rsid' in row and row.rsid.dtype == tstr:
+        names_elems.append("rsid")
+    names_col = fx.expr(
+        f"filter(nullif(array({','.join(names_elems)}), array()), n -> isnotnull(n))").alias("names")
+
+    reference_allele_col = fx.element_at("alleles", 1).alias("referenceAllele")
+
+    alternate_alleles_col = fx.expr("slice(alleles, 2, size(alleles) - 1)").alias("alternateAlleles")
+
+    base_cols = [
+        contig_name_col, start_col, end_col, names_col, reference_allele_col, alternate_alleles_col
+    ]
+    assert check_return_type(base_cols)
+    return base_cols
+
+
+def __get_other_cols(row: StructExpression) -> List[Column]:
+    assert check_argument_types()
+
+    other_cols = []
+    if 'cm_position' in row and row.cm_position.dtype == tfloat64:
+        other_cols.append(fx.col("cm_position").alias("position"))
+    if 'qual' in row and row.qual.dtype == tfloat64:
+        # -10 qual means missing
+        other_cols.append(fx.expr("if(qual = -10, null, qual)").alias("qual"))
+    # null filters means missing, [] filters means PASS
+    if 'filters' in row and row.filters.dtype == tset(tstr):
+        other_cols.append(
+            fx.expr("if(size(filters) = 0, array('PASS'), if(isnull(filters), array(), filters))").
+            alias("filters"))
+    # Rename info.* columns to INFO_*
+    if 'info' in row and isinstance(row.info.dtype, tstruct):
+        for f in row.info:
+            if row.info[f].dtype == tbool:
+                # FLAG INFO fields should be null if not present; Hail encodes them as false
+                other_cols.append(
+                    fx.expr(f"if(`info.{f}` = false, null, `info.{f}`)").alias(f"INFO_{f}"))
+            else:
+                other_cols.append(fx.col(f"`info.{f}`").alias(f"INFO_{f}"))
+
+    assert check_return_type(other_cols)
+    return other_cols
+
+
+def __require_row_variant_w_struct_locus(mt: MatrixTable) -> NoReturn:
```

Comment:
Why should we require these fields?

Reply:
I'm not sure if there is any MatrixTable without these fields that we could meaningfully convert to a Glow-compatible Spark DataFrame; this quickly signals that the input is misconfigured.